text/0003-templates.md

```diff
@@ -0,0 +1,258 @@
+
+- In projects that bundle dependencies, multiple versions inflate the size of the final result deployed to users.
+- Differing versions result in multiple copies that may not interact well at runtime, especially if features like [`Symbol()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) are used. For example, [React hooks will error if a component is rendered with a different copy of React in the same app](https://reactjs.org/warnings/invalid-hook-call-warning.html#duplicate-react).
+- For TypeScript, multiple copies of the same `@types` package causes compile errors from mismatching definitions. The compiler diagnostic for this is usually: *"argument of type `Foo` is not assignable to parameter of type `Foo`"*. For developers that have seen this before, they may realize this diagnostic is due to a dependency version mismatch. For developers new to TypeScript, *"`Foo` is not assignable to `Foo`"* is very confusing.
+
+While there are situations differing versions are intentional, this is more often accidental. Multiple differing versions arise from not reviewing `pnpm-lock.yaml` file changes or not searching for existing dependency specifiers before adding a new one. The later is typically unwritten convention in most monorepos.
+
+### Authoring Fields
+
+Fields such as `author`, `license`, and `repository` are typically expected to be the same across all packages in a monorepo. It would be easier to set this in a singular source of truth rather than every package.
+
+## Detailed Explanation
+
+A "*Template*" is defined as a normal `package.json` file with the `pnpm.template` field set.
+
+```json5
+{
+  "name": "@example/react-template",
+  "pnpm": {
+    "template": true,
+  },
+  "version": "0.1.0",
+  "author": "Example Team <team@team.example>",
+  "license": "MIT",
+  "scripts": {
+    "compile": "tsc"
+  },
+  "dependencies": {
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+  }
+}
+```
+
+Templates must opt into being a template by defining `pnpm.template`. This is a safety mechanism to:
+
+1. Intentionally prevent existing packages from being referred to for that purpose. Templates have different obligations and usage than standard packages that authors of templates should be aware of.
+2. Allow checks to be performed on the template before publishing and during consumption. Not all `package.json` fields may be valid on a template.
+
+A package can reference the template to populate different portions of its definition. The specific feature is determined by `pnpm.templates`: `extends`, `toolkit`, and `catalog`.
+
+### Extends
+
+Using the `pnpm.templates.extends` mechanism, `package.json` fields from templates are copied with small exceptions. The `name`, `dist`, and underscore prefixed fields (e.g. `_npmUser`) are not copied since they're typically specific to the template package itself.
```

Comment:
I'm also still unsure if `extends` should inherit all `package.json` fields by default, or only inherit a filtered list hard-coded into pnpm.

Reply:
I think all is OK. Although, it makes me wonder. If templates will extend packageExtensions and the likes, then such templates could be useful in the open source. Like at the moment pnpm ships with some packageExtensions to fix known issues in the ecosystem. If people will publish such open source package, then the name, license, repository fields of such packages will not have to be used for extending.

Maybe it makes sense to use a separate somehow fields that are meant to be used for extending.

Reply:
> I think all is OK. Although, it makes me wonder. If templates will extend packageExtensions and the likes, then such templates could be useful in the open source. Like at the moment pnpm ships with some packageExtensions to fix known issues in the ecosystem.

I'm also excited to see the potential usefulness of this in open source. 🙂

> If people will publish such open source package, then the name, license, repository fields of such packages will not have to be used for extending.

Going to think about this more, but perhaps the concept of `public`/`private` modifiers is necessary.

Reply:
Agreed with "all" being OK. Different templates could extend different concerns. If templates can be composed via `extends` then we have what we need.

Reply:
For now we'll avoid an `extends` template.

- Rationale in https://github.com/pnpm/rfcs/pull/3#issuecomment-1493167623
- Change pushed in commit c1f3f2aa6c8120c495e4a3ced884dcb445e54085.

Going to hit the "_Resolve conversation_" button for this thread.